### PR TITLE
import into: make post process a step to be able cancel

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #42930

Problem Summary:

### What is changed and how it works?
- make Scheduler/MinimalTaskExecutor a pair for each step
- make post-process a step so we can cancel it using existing cancel mechanism
- set to null when start/end time is 0

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/importinto/dispatcher.go [hunk 1]
```diff
@@ -205,29 +210,29 @@ func (h *flowHandle) ProcessNormalFlow(ctx context.Context, handle dispatcher.Ta
...
+		gTask.Step = StepImport
 		return metaBytes, nil
-	case Import:
-		defer func() {
-			if err == nil {
-				err = finishJob(ctx, handle, taskMeta, &importer.JobSummary{ImportedRows: taskMeta.Result.LoadedRowCnt})
-			} else {
-				// todo: we're not running in a transaction with task update, there might be case
-				// failJob return error, but task update succeed.
-				if err2 := failJob(ctx, handle, taskMeta, err.Error()); err2 != nil {
```
Comment (zimulala) [inline disttask/importinto/dispatcher.go:217]: Why do we remove this logic?

File: disttask/importinto/proto.go [hunk 2]
```diff
@@ -25,10 +25,15 @@ import (
...
+// Steps of IMPORT INTO, each step is represented by one or multiple subtasks.
+// the initial step is StepInit(-1)
+// steps are processed in the following order: StepInit -> StepImport -> StepPostProcess
 const (
-	// Import we sort source data and ingest it into TiKV in this step.
-	Import int64 = 1
+	// StepImport we sort source data and ingest it into TiKV in this step.
+	StepImport int64 = 1
+	// StepPostProcess we verify checksum and add index in this step.
+	// todo: might split into StepValidate and StepAddIndex later.
```
Comment (zimulala) [inline disttask/importinto/proto.go:35]:
```suggestion
	// TODO: Might split into StepValidate and StepAddIndex later.
```

File: disttask/importinto/dispatcher.go (hunk 1 above)
Comment (D3Hunter) [reply disttask/importinto/dispatcher.go:217]: i thought framework will call ProcessErrFlow and handle it there previously, will change it

Comment (D3Hunter) [reply disttask/importinto/dispatcher.go:217]: https://github.com/pingcap/tidb/pull/44616/commits/8b4e8f3ba4f2e7a679f02347e09f68178395ea51

Comment (zimulala) [reply disttask/importinto/dispatcher.go:217]: Got it.

File: disttask/framework/scheduler/scheduler.go [hunk 3]
```diff
@@ -307,17 +308,19 @@ func (s *InternalSchedulerImpl) Rollback(ctx context.Context, task *proto.Task)
 }
 
 func createScheduler(task *proto.Task) (Scheduler, error) {
-	constructor, ok := schedulerConstructors[task.Type]
+	key := getKey(task.Type, task.Step)
+	constructor, ok := schedulerConstructors[key]
 	if !ok {
-		return nil, errors.Errorf("constructor of scheduler for type %s not found", task.Type)
+		return nil, errors.Errorf("ctor of scheduler for key %s not found", key)
```
Comment (GMHDBJD) [inline disttask/framework/scheduler/scheduler.go:314]: constructor

File: disttask/framework/scheduler/scheduler.go [hunk 4]
```diff
@@ -307,17 +308,19 @@ func (s *InternalSchedulerImpl) Rollback(ctx context.Context, task *proto.Task)
...
 	return constructor(task.ID, task.Meta, task.Step)
 }
 
 func createSubtaskExecutor(minimalTask proto.MinimalTask, tp string, step int64) (SubtaskExecutor, error) {
-	constructor, ok := subtaskExecutorConstructors[tp]
+	key := getKey(tp, step)
+	constructor, ok := subtaskExecutorConstructors[key]
 	if !ok {
-		return nil, errors.Errorf("constructor of subtask executor for type %s not found", tp)
+		return nil, errors.Errorf("ctor of subtask executor for key %s not found", key)
```
Comment (GMHDBJD) [inline disttask/framework/scheduler/scheduler.go:323]: ditto

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: disttask/importinto/dispatcher.go [hunk 5]
```diff
@@ -317,53 +331,37 @@ func preProcess(ctx context.Context, handle dispatcher.TaskHandle, gTask *proto.
 }
 
 // postProcess does the post-processing for the task.
-func postProcess(ctx context.Context, handle dispatcher.TaskHandle, gTask *proto.Task, taskMeta *TaskMeta, logger *zap.Logger) (_ []*SubtaskMeta, err error) {
+func postProcess(ctx context.Context, taskMeta *TaskMeta, subtaskMeta *PostProcessStepMeta, logger *zap.Logger) (err error) {
```
Comment (GMHDBJD) [inline disttask/importinto/dispatcher.go:334]: move to scheduler？

File: executor/import_into.go [hunk 6]
```diff
@@ -164,7 +164,7 @@ func (e *ImportIntoExec) fillJobInfo(ctx context.Context, jobID int64, req *chun
 	}); err != nil {
 		return err
 	}
-	fillOneImportJobInfo(info, req, 0)
+	fillOneImportJobInfo(info, req, -1)
```
Comment (zimulala) [inline executor/import_into.go:167]: Could we add a constant for `-1`?

File: disttask/framework/scheduler/scheduler.go (hunk 3 above)
Comment (D3Hunter) [reply disttask/framework/scheduler/scheduler.go:314]: abbr of constructor

File: disttask/importinto/scheduler.go [hunk 7]
```diff
@@ -179,37 +176,80 @@ func (s *ImportScheduler) OnSubtaskFinished(ctx context.Context, subtaskMetaByte
...
-		proto.ImportInto,
+	prepareFn := func(taskID int64, bs []byte, step int64) (*TaskMeta, *zap.Logger, error) {
+		taskMeta := TaskMeta{}
+		if err := json.Unmarshal(bs, &taskMeta); err != nil {
+			return nil, nil, err
+		}
+		logger := logutil.BgLogger().With(
+			zap.String("component", "scheduler"),
+			zap.String("type", proto.ImportInto),
+			zap.Int64("task-id", taskID),
```
Comment (zimulala) [inline disttask/importinto/scheduler.go:221]:
```suggestion
			zap.Int64("task-ID", taskID),
```

File: disttask/importinto/dispatcher.go [hunk 8]
```diff
@@ -176,13 +176,34 @@ func (h *flowHandle) unregisterTask(ctx context.Context, task *proto.Task) {
...
 		return nil, err
 	}
-	logger.Info("process normal flow", zap.Any("task_meta", taskMeta), zap.Any("step", gTask.Step))
+	logger.Info("process normal flow", zap.Any("task_meta", taskMeta))
+
+	var taskFinished bool
+	defer func() {
+		if taskFinished {
+			// todo: we're not running in a transaction with task update
+			if err2 := h.finishJob(ctx, handle, gTask, taskMeta, logger); err2 != nil {
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply disttask/importinto/dispatcher.go:196]: `finishJob` fails when execute sql, might fail again on `failJob`, now it's not run in the same transaction with framework task, might cause state in-consistency.

File: disttask/importinto/dispatcher.go [hunk 9]
```diff
@@ -176,13 +176,34 @@ func (h *flowHandle) unregisterTask(ctx context.Context, task *proto.Task) {
...
+				err = err2
+			}
+		} else if err != nil && !h.IsRetryableErr(err) {
+			if err2 := h.failJob(ctx, handle, gTask, taskMeta, logger, err.Error()); err2 != nil {
+				// todo: we're not running in a transaction with task update, there might be case
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply disttask/importinto/dispatcher.go:201]: i think the case of `todo` doesn't matter, will leave it

File: disttask/importinto/scheduler.go (hunk 7 above)
Comment (D3Hunter) [reply disttask/importinto/scheduler.go:221]: it works like a json field name, better leave it lower case

File: disttask/importinto/subtask_executor.go [hunk 10]
```diff
@@ -32,13 +32,13 @@ var TestSyncChan = make(chan struct{})
 
 // ImportMinimalTaskExecutor is a minimal task executor for IMPORT INTO.
 type ImportMinimalTaskExecutor struct {
-	task *MinimalTaskMeta
+	mTtask *importStepMinimalTask
```
Comment (GMHDBJD) [inline disttask/importinto/subtask_executor.go:35]: mTask?

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: disttask/importinto/subtask_executor.go (hunk 10 above)
Comment (D3Hunter) [reply disttask/importinto/subtask_executor.go:35]: `m`inimal `Task`

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**